src/str.rs

```diff
@@ -122,160 +122,236 @@ pub(crate) fn fmt_scientific_notation(
 }
 
 // dedicated implementation for the most common case.
+#[inline]
 pub(crate) fn parse_str_radix_10(str: &str) -> Result<Decimal, crate::Error> {
-    if str.is_empty() {
-        return Err(Error::from("Invalid decimal: empty"));
+    let bytes = str.as_bytes();
+    // handle the sign
+
+    if bytes.len() < BYTES_TO_OVERFLOW_U64 {
+        parse_str_radix_10_dispatch::<false>(bytes)
+    } else {
+        parse_str_radix_10_dispatch::<true>(bytes)
     }
+}
 
-    let mut offset = 0;
-    let mut len = str.len();
-    let bytes = str.as_bytes();
-    let mut negative = false; // assume positive
+#[inline]
+fn parse_str_radix_10_dispatch<const BIG: bool>(bytes: &[u8]) -> Result<Decimal, crate::Error> {
+    match bytes {
+        [b, rest @ ..] => byte_dispatch_u64::<false, false, false, BIG, true>(rest, 0, 0, *b),
+        [] => tail_error("Invalid decimal: empty"),
+    }
+}
 
-    // handle the sign
-    if bytes[offset] == b'-' {
-        negative = true; // leading minus means negative
-        offset += 1;
-        len -= 1;
-    } else if bytes[offset] == b'+' {
-        // leading + allowed
-        offset += 1;
-        len -= 1;
+#[inline]
+fn overflow_64(val: u64) -> bool {
+    val >= WILL_OVERFLOW_U64
+}
+
+#[inline]
+pub fn overflow_128(val: u128) -> bool {
+    val >= OVERFLOW_U96
+}
+
+#[inline]
+fn dispatch_next<const POINT: bool, const NEG: bool, const HAS: bool, const BIG: bool>(
+    bytes: &[u8],
+    data64: u64,
+    scale: u8,
+) -> Result<Decimal, crate::Error> {
+    if let Some((next, bytes)) = bytes.split_first() {
+        byte_dispatch_u64::<POINT, NEG, HAS, BIG, false>(bytes, data64, scale, *next)
+    } else {
+        handle_data::<NEG, HAS>(data64 as u128, scale)
     }
+}
 
-    // should now be at numeric part of the significand
-    let mut digits_before_dot: i32 = -1; // digits before '.', -1 if no '.'
-    let mut coeff = ArrayVec::<_, MAX_STR_BUFFER_SIZE>::new(); // integer significand array
+#[inline(never)]
+fn non_digit_dispatch_u64<const POINT: bool, const NEG: bool, const HAS: bool, const BIG: bool, const FIRST: bool>(
+    bytes: &[u8],
+    data64: u64,
+    scale: u8,
+    b: u8,
+) -> Result<Decimal, crate::Error> {
+    match b {
+        b'-' if FIRST && !HAS => dispatch_next::<false, true, false, BIG>(bytes, data64, scale),
+        b'+' if FIRST && !HAS => dispatch_next::<false, false, false, BIG>(bytes, data64, scale),
+        b'_' if HAS => handle_separator::<POINT, NEG, BIG>(bytes, data64, scale),
+        b => tail_invalid_digit(b),
+    }
+}
 
-    let mut maybe_round = false;
-    while len > 0 {
-        let b = bytes[offset];
-        match b {
-            b'0'..=b'9' => {
-                coeff.push(u32::from(b - b'0'));
-                offset += 1;
-                len -= 1;
+#[inline]
+fn byte_dispatch_u64<const POINT: bool, const NEG: bool, const HAS: bool, const BIG: bool, const FIRST: bool>(
```

Comment:
My one qualm with these const generics is perhaps knowing what each does. Even though this is an internal method it'd be great to add a little bit of documentation just to cover what the purpose of each is for. 
* `POINT` - a decimal point has been seen
* `NEG` - we've encountered a `-` and the number is negative
* `HAS` - a digit has been encountered (when `HAS` is `false` it's invalid)
* `BIG` - a number that uses 96 bits instead of only 64 bits
* `FIRST` - true if it is the first byte in the string